mava/systems/q_learning/sebulba/rec_iql.py

```diff
@@ -0,0 +1,782 @@
+
+    def learner_fn(
+        learner_state: LearnerState, traj_batch: Transition
+    ) -> Tuple[LearnerState, Metrics]:
+        """Learner function.
+
+        This function represents the learner, it updates the network parameters
+        by iteratively applying the `_update_step` function for a fixed number of
+        updates. The `_update_step` function is vectorized across learner devices.
+
+        Args:
+            learner_state (NamedTuple):
+                - params (Params): The initial model parameters.
+                - opt_states (OptStates): The initial optimizer state.
+                - step_counter int): Number of learning steps.
+            traj_batch (Transition): The collected trainig data.
+        """
+        learner_state, loss_info = _update_step(learner_state, traj_batch)
+
+        return learner_state, loss_info
+
+    return learner_fn
+
+
+def learner_thread(
+    learn_fn: SebulbaLearnerFn[LearnerState, Transition],
+    learner_state: LearnerState,
+    config: DictConfig,
+    eval_queue: Queue,
+    pipeline: Pipeline,
+    params_sources: Sequence[ParamsSource],
+) -> None:
+    for _ in range(config.arch.num_evaluation):
+        # Create the lists to store metrics and timings for this learning iteration.
+        ep_metrics: List[Dict] = []
+        train_metrics: List[Dict] = []
+        rollout_times: List[Dict] = []
+        learn_times: Dict[str, List[float]] = defaultdict(list)
+
+        with RecordTimeTo(learn_times["learner_time_per_eval"]):
+            for _ in range(config.system.num_updates_per_eval):
+                # Get the trajectory batch from the pipeline
+                # This is blocking so it will wait until the pipeline has data.
+                with RecordTimeTo(learn_times["rollout_get_time"]):
+                    traj_batch, (rollout_time, ep_metric) = pipeline.get()
+                # Update the networks
+                with RecordTimeTo(learn_times["learning_time"]):
+                    learner_state, train_metric = learn_fn(learner_state, traj_batch)
+
+                train_metrics.append(train_metric)
+                if ep_metric is not None:
+                    ep_metrics.append(ep_metric)
+                    rollout_times.append(rollout_time)
+
+                # Update all the params sources so all actors can get the latest params
+                params = jax.block_until_ready(learner_state.params)
+                for source in params_sources:
+                    source.update(params.online)
+
+        # Pass all the metrics and  params to the main thread (evaluator) for logging and evaluation
+        if ep_metrics:
+            ep_metrics = tree.map(lambda *x: np.asarray(x), *ep_metrics)
+            train_metrics = tree.map(lambda *x: np.asarray(x), *train_metrics)
+
+        timing_dict = tree.map(lambda *x: np.mean(x), *rollout_times) | learn_times
+        timing_dict = tree.map(np.mean, timing_dict, is_leaf=lambda x: isinstance(x, list))
```

Comment:
2 things, can we call this time_metrics and can you add a shape explainer comment as it's a bit hard to work out what is happening here